command/agent/command.go

```diff
@@ -978,66 +978,31 @@ func (c *Command) handleRetryJoin(config *Config) error {
 	return nil
 }
 
-// handleSignals blocks until we get an exit-causing signal
-func (c *Command) handleSignals() int {
-	signalCh := make(chan os.Signal, 4)
-	signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM, syscall.SIGHUP, syscall.SIGPIPE)
-
-	// Signal readiness only once signal handlers are setup
-	sdSock, err := openNotify()
-	if err != nil {
-		c.agent.logger.Debug("notify socket could not be accessed", "error", err)
-	}
-	if sdSock != nil {
-		defer sdSock.Close()
-	}
-	sdNotify(sdSock, sdReady)
-
-	// Wait for a signal
-WAIT:
-	var sig os.Signal
-	select {
-	case s := <-signalCh:
-		sig = s
-	case <-winsvc.ShutdownChannel():
-		sig = os.Interrupt
-	case <-c.ShutdownCh:
-		sig = os.Interrupt
-	case <-c.retryJoinErrCh:
-		return 1
-	}
+// terminateGracefully attempts a graceful leave
+func (c *Command) terminateGracefully(signalCh chan os.Signal, sdSock io.Writer) int {
+	sdNotify(sdSock, sdStopping)
 
-	// Skip any SIGPIPE signal and don't try to log it (See issues #1798, #3554)
-	if sig == syscall.SIGPIPE {
-		goto WAIT
-	}
+	gracefulCh := make(chan struct{})
+	defer close(gracefulCh)
 
-	c.Ui.Output(fmt.Sprintf("Caught signal: %v", sig))
+	timeout := gracefulTimeout
 
-	// Check if this is a SIGHUP
-	if sig == syscall.SIGHUP {
-		sdNotifyReloading(sdSock)
-		c.handleReload()
-		sdNotify(sdSock, sdReady)
-		goto WAIT
+	config := c.agent.config
```

Comment:
We should be using the `configLock` around the agent configuration here. And then we probably need to release it right away rather than waiting for the `defer`.

Reply:
changing to the client config takes care of this